components/brave_perf_predictor/browser/bandwidth_savings_predictor.cc

```diff
@@ -0,0 +1,151 @@
+  if (entity_name.has_value())
+    feature_map_["thirdParties." + entity_name.value() + ".blocked"] = 1;
+}
+
+void BandwidthSavingsPredictor::OnResourceLoadComplete(
+    const GURL& main_frame_url,
+    const content::mojom::ResourceLoadInfo& resource_load_info) {
+  // If the resource load info comes without a valid corresponding
+  // main frame URL, ignore it
+  if (main_frame_url.is_empty() || !main_frame_url.has_host() ||
+      !main_frame_url.SchemeIsHTTPOrHTTPS()) {
+    return;
+  }
+  main_frame_url_ = main_frame_url;
+
+  bool is_third_party = !net::registry_controlled_domains::SameDomainOrHost(
+      main_frame_url, resource_load_info.url,
+      net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  if (is_third_party) {
+    feature_map_["resources.third-party.requestCount"] += 1;
+    feature_map_["resources.third-party.size"] +=
+        resource_load_info.raw_body_bytes;
+  }
+
+  feature_map_["resources.total.requestCount"] += 1;
+  feature_map_["resources.total.size"] += resource_load_info.raw_body_bytes;
+  feature_map_["transfer.total.size"] +=
+      resource_load_info.total_received_bytes;
+  std::string resource_type;
+  switch (resource_load_info.resource_type) {
+    case content::ResourceType::kMainFrame:
+      resource_type = "document";
+      break;
+    case content::ResourceType::kSubFrame:
+      resource_type = "document";
+      break;
+    case content::ResourceType::kStylesheet:
+      resource_type = "stylesheet";
+      break;
+    case content::ResourceType::kScript:
+      resource_type = "script";
+      break;
+    case content::ResourceType::kImage:
+      resource_type = "image";
+      break;
+    case content::ResourceType::kFontResource:
+      resource_type = "font";
+      break;
+    case content::ResourceType::kMedia:
+      resource_type = "media";
+      break;
+    default:
+      resource_type = "other";
+      break;
+  }
+  feature_map_["resources." + resource_type + ".requestCount"] += 1;
+  feature_map_["resources." + resource_type + ".size"] +=
+      resource_load_info.raw_body_bytes;
+}
+
+double BandwidthSavingsPredictor::Predict() {
+  if (main_frame_url_.is_empty() || !main_frame_url_.has_host() ||
+      !main_frame_url_.SchemeIsHTTPOrHTTPS()) {
+    Reset();
+    return 0;
+  }
+  if (feature_map_["transfer.total.size"] > 0) {
+    VLOG(2) << main_frame_url_ << " total download size "
+            << feature_map_["transfer.total.size"] << " bytes";
+  }
+
+  // Short-circuit if nothing got blocked
+  if (feature_map_["adblockRequests"] < 1) {
+    Reset();
+    return 0;
+  }
+  if (VLOG_IS_ON(3)) {
+    VLOG(3) << "Predicting on feature map:";
+    auto it = feature_map_.begin();
+    while (it != feature_map_.end()) {
+      VLOG(3) << it->first << " :: " << it->second;
+      it++;
+    }
+  }
+  double prediction = ::brave_perf_predictor::Predict(feature_map_);
+  VLOG(2) << main_frame_url_ << " estimated saving " << prediction << " bytes";
+  Reset();
```

Comment:
wow, this was out of the blue sky... This behavior should be definitely mentioned in the comment, or even removed since  the whole thing is quite misleading. This is rather `PredictAndReset()` or `GetSavedBytesCountAndReset()`